File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/ValidatingOngoingThrowingClause.java
package fr.xgouchet.mesmaeker.mockito;

import java.util.function.Consumer;

import static fr.xgouchet.mesmaeker.core.ValidatingUtils.verifyConsumerThrows;

/**
 * @author Xavier F. Gouchet
 */
public class ValidatingOngoingThrowingClause<T> implements OngoingThrowingClause<T> {


    private final T subject;
    private final Class<? extends Throwable>[] expectedThrowables;


    public ValidatingOngoingThrowingClause(T subject, Class<? extends Throwable>... toBeThrown) {
        this.subject = subject;
        expectedThrowables = toBeThrown;
    }

    public ValidatingOngoingThrowingClause(T subject, Throwable... toBeThrown) {
        this.subject = subject;
        //noinspection unchecked
        expectedThrowables = new Class[toBeThrown.length];

        for (int i = 0; i < toBeThrown.length; i++) {
            expectedThrowables[i] = toBeThrown[i].getClass();
        }
    }

    @Override public OngoingThrowingClause<T> when(Consumer<T> consumer) {

        for (Class<? extends Throwable> t : expectedThrowables) {
            verifyConsumerThrows(subject, consumer, t);
        }
        return this;
    }
}

File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/ValidatingOngoingThrowingClause.kt
package fr.xgouchet.mesmaeker.mockito

import fr.xgouchet.mesmaeker.core.ValidatingUtils.verifyBlockThrows

/**
 * @author Xavier F. Gouchet
 */
class ValidatingOngoingThrowingClause<T> : OngoingThrowingClause<T> {


    private val subject: T
    private val expectedThrowables: List<Class<out Throwable>>


    @SafeVarargs
    constructor(subject: T, vararg toBeThrown: Class<out Throwable>) {
        this.subject = subject
        expectedThrowables = listOf(*toBeThrown)
    }

    constructor(subject: T, vararg toBeThrown: Throwable) {
        this.subject = subject

        expectedThrowables = listOf(*toBeThrown)
                .map { it.javaClass }
    }

    override fun `when`(block: (T) -> Unit): OngoingThrowingClause<T> {
        return whenever(block)
    }

    override fun whenever(block: (T) -> Unit): OngoingThrowingClause<T> {
        for (t in expectedThrowables) {
            verifyBlockThrows(subject, block, t)
        }
        return this
    }
}
